src/roswire/ros2/reader.py

```diff
@@ -0,0 +1,129 @@
+        argv: Sequence[str], optional
+            An optional sequence of command-line arguments that should be
+            supplied to :code:`roslaunch`.
+
+        Raises
+        ------
+        LaunchFileNotFound
+            If the given launch file could not be found in the package.
+        """
+        logger.debug("Copying launch extraction script")
+        files = self._app_instance.files
+        host_script = pkg_resources.resource_filename(
+            'roswire',
+            'resources/launch_extractor._py')
+        files.copy_from_host(host_script,
+                             '/launch_extractor.py')
+
+        config_nodes = self._process_launch_on_app_instance(fn, files)
+        lc = self._read_launch_config_from_dict(config_nodes)
+        return lc
+
+    def _read_launch_config_from_dict(self,
+                                      config_nodes: Sequence[Dict[str, Any]],
+                                      ) -> LaunchConfig:
+        cfg = LaunchConfig()
+        cfg = self._load_launch_objects(cfg,
+                                        [list(config_nodes)])
+        logger.debug(f'launch configuration: {cfg}')
+        return cfg
+
+    def _process_launch_on_app_instance(self,
+                                        filename: str,
+                                        files: Any
+                                        ) -> Sequence[Dict[str, Any]]:
+        assert self._app_instance is not None
+        output = shlex.quote(os.path.basename(filename) + '.json')
+        cmd = f'python3 /launch_extractor.py --output' \
+              f' {output} {shlex.quote(filename)}'
+        logger.debug(f"Running the script in the container: {cmd}")
+        self._app_instance.shell.check_call(cmd)
+        logger.debug(f"Reading {output} on container")
+        config_json = files.read(output)
+        config_nodes = json.loads(config_json)
+        return config_nodes
+
+    def _load_launch_objects(self,
+                             cfg: LaunchConfig,
+                             node_list: Sequence[Sequence[Dict[str, Any]]]
+                             ) -> LaunchConfig:
+        for nodes in node_list:
+            for node in nodes:
+                if node['__TYPE__'] == 'Node':
+                    nc = self._read_node_from_dict(node)
+                    cfg = cfg.with_node(nc)
+                else:
+                    # The __TYPE__ isn't known (this is for futureproofing)
+                    raise NotImplementedError
+        return cfg
+
+    def _read_node_from_dict(self, node: Dict[str, Any]) -> NodeConfig:
+        args = ' '.join(node.get('args', []))
+        remappings = tuple(node.get('remappings', []))
+        nc = NodeConfig(
+            name=node['name'],
+            namespace=node['namespace'],
+            package=node['package'],
+            executable_path=node['executable_path'],
+            executable_type=ExecutableType[
+                node['executable_type']
```

Comment:
Separate line makes this look like a list from a glance.

Reply:
Done